lib/identity.go

```diff
@@ -130,7 +131,15 @@ func (i *Identity) RegisterAndEnroll(req *api.RegistrationRequest) (*Identity, e
 func (i *Identity) Reenroll(req *api.ReenrollmentRequest) (*EnrollmentResponse, error) {
 	log.Debugf("Reenrolling %s", util.StructToString(req))
 
-	csrPEM, key, err := i.client.GenCSR(req.CSR, i.GetName())
+	var key bccsp.Key
+	if req.CSR.KeyRequest.ReuseKey {
+		val, err := i.GetX509Credential().Val()
+		if err != nil {
+			return nil, err
+		}
+		key = val.(*x509.Signer).Key()
+	}
+	csrPEM, key, err := i.client.GenCSR(req.CSR, i.GetName(), key)
```

Comment:
Here is the meat of the change. Glancing at the code, I see `Reenroll` checking a nested struct in `CSRInfo` (no nil check - which goes to the comment I made in main_test) and using that information to extract a key based on `ReuseKey`. This structure seems to be motivating the change to `GenCSR` (breaking client API) to pass down the key.

My guess is that this PR would have a much smaller surface area and much smaller impact if a new interface function was introduced on the client to handle this re-enrollment flow with an existing key.

Reply:
Done